File: atms/go/funify.go
// -*- Mode: Go; -*-

// Extra pattern-matching facilities for FTRE
// Open-coding unification for rule compilation
// Last edited: 1/29/93, KDF

// Copyright (c) 1988-1992, Kenneth D. Forbus and Johan de Kleer,
// All Rights Reserved.

// See the file legal.txt for a paragraph stating scope of permission
// and disclaimer of warranty. The above copyright notice and that
// paragraph must be included in any separate copy of this file.

package atms

import "fmt"

// Symbol represents a Lisp symbol as a Go string constant.
// Symbols like QUOTE, CONS, EQUAL, NULL, CONSP, AND, CAR, CDR, NUMBERP
// are used in generated test expressions.
type Symbol string

const (
	SymQuote   Symbol = "QUOTE"
	SymCons    Symbol = "CONS"
	SymEqual   Symbol = "EQUAL"
	SymNull    Symbol = "NULL"
	SymConsp   Symbol = "CONSP"
	SymAnd     Symbol = "AND"
	SymCar     Symbol = "CAR"
	SymCdr     Symbol = "CDR"
	SymNumberp Symbol = "NUMBERP"
	SymEval    Symbol = "EVAL"
)

// Quotize transforms a pattern into an s-expression that, when evaluated,
// reconstructs the pattern. Variables are left as-is, non-list atoms
// are quoted, :EVAL forms are unwrapped, and lists become nested CONS forms.
func Quotize(pattern interface{}) interface{} {
	if pattern == nil {
		return nil
	}
	if IsVariable(pattern) {
		return pattern
	}
	cons, ok := pattern.(*Cons)
	if !ok {
		// Non-list atom: (QUOTE pattern)
		return SExprList(SymQuote, pattern)
	}
	// Check for :EVAL
	if sym, ok := cons.Car.(Symbol); ok && sym == SymEval {
		// (cadr pattern)
		return Car(cons.Cdr)
	}
	// (CONS (quotize (car pattern)) (quotize (cdr pattern)))
	return SExprList(SymCons, Quotize(cons.Car), Quotize(cons.Cdr))
}

// RLetClause represents a single variable binding in an rlet form.
type RLetClause struct {
	Var  string
	Expr interface{}
}

// RLetExpand expands an rlet form at runtime. It processes variable specs,
// calling Quotize on non-:EVAL values, and expands the body with the
// bound variables in scope.
// boundVars is the current set of bound variable names.
// varSpecs is a list of (variable, expression) pairs.
// bodyExpander is a function that expands the body given updated bound vars.
func RLetExpand(boundVars []string, varSpecs []RLetClause, bodyExpander func(boundVars []string) interface{}) interface{} {
	// Extend bound vars with the new variable names.
	newBound := make([]string, len(boundVars), len(boundVars)+len(varSpecs))
	copy(newBound, boundVars)
	for _, spec := range varSpecs {
		newBound = append(newBound, spec.Var)
	}

	// Process each let clause.
	processedSpecs := make([]RLetClause, len(varSpecs))
	for i, spec := range varSpecs {
		expr := spec.Expr
		// Check if expression is (:EVAL ...)
		if cons, ok := expr.(*Cons); ok {
			if sym, ok := cons.Car.(Symbol); ok && sym == SymEval {
				// Use the raw expression (cadr)
				expr = Car(cons.Cdr)
			} else {
				expr = Quotize(expr)
			}
		} else if expr != nil && !IsVariable(expr) {
			expr = Quotize(expr)
		}
		processedSpecs[i] = RLetClause{Var: spec.Var, Expr: expr}
	}

	// Build the let form as an s-expression.
	// (LET ((var1 expr1) (var2 expr2) ...) body...)
	body := bodyExpander(newBound)
	letBindings := make([]interface{}, len(processedSpecs))
	for i, spec := range processedSpecs {
		letBindings[i] = SExprList(spec.Var, spec.Expr)
	}
	return SExprList(Symbol("LET"), SliceToSExpr(letBindings), body)
}

// PatternFreeVariables returns the free variables in a pattern,
// given the current set of bound variables.
func PatternFreeVariables(pattern interface{}, boundVars []string) []string {
	return patternFreeVars1(pattern, nil, boundVars)
}

// patternFreeVars1 accumulates free variables found in pattern.
func patternFreeVars1(pattern interface{}, vars []string, boundVars []string) []string {
	if pattern == nil {
		return vars
	}
	if IsVariable(pattern) {
		s := pattern.(string)
		if stringMember(s, vars) || stringMember(s, boundVars) {
			return vars
		}
		return append(vars, s)
	}
	cons, ok := pattern.(*Cons)
	if !ok {
		// Atom (not a variable, not a list) => no free vars.
		return vars
	}
	vars = patternFreeVars1(cons.Car, vars, boundVars)
	return patternFreeVars1(cons.Cdr, vars, boundVars)
}

// MatchBodyResult holds the results of GenerateMatchBody.
type MatchBodyResult struct {
	Tests        []interface{}
	BindingSpecs []interface{}
}

// GenerateMatchBody generates the test body for pattern matching.
// pattern is the pattern to match against, vars is the list of pattern
// variables, and extraTest is an optional additional test expression.
// boundVars is needed for checking free variables in the extra test.
func GenerateMatchBody(pattern interface{}, vars []string, extraTest interface{}, boundVars []string) (*MatchBodyResult, error) {
	var structureTests []interface{}
	var varAlist []Binding
	var equalTests []interface{}
	var bindingSpecs []interface{}

	tests := GenerateUnifyTests(pattern, vars, nil, Symbol("P"))
	for _, rawTest := range SExprToSlice(tests) {
		testCons, ok := rawTest.(*Cons)
		if !ok {
			continue
		}
		first := testCons.Car
		if IsVariable(first) {
			// Test looks like (?x (nth p) (nth p) ...)
			rest := SExprToSlice(testCons.Cdr)
			pairwise := GeneratePairwiseTests(testCons.Cdr)
			for _, pt := range SExprToSlice(pairwise) {
				equalTests = append(equalTests, pt)
			}
			if extraTest != nil {
				lastElem := rest[len(rest)-1]
				varAlist = append(varAlist, Binding{Var: first.(string), Val: lastElem})
			}
			lastElem := rest[len(rest)-1]
			bindingSpecs = append(bindingSpecs, lastElem)
		} else {
			structureTests = append(structureTests, rawTest)
		}
	}

	// Substitute variable alist into extra test.
	extraTest = Sublis(varAlist, extraTest)

	// Check for free variables in extra test.
	if extraTest != nil {
		freeVars := PatternFreeVariables(extraTest, boundVars)
		if len(freeVars) > 0 {
			return nil, fmt.Errorf("rule test includes free variable: %v", freeVars)
		}
	}

	// Combine all tests.
	var allTests []interface{}
	allTests = append(allTests, structureTests...)
	allTests = append(allTests, equalTests...)
	if extraTest != nil {
		allTests = append(allTests, extraTest)
	}

	return &MatchBodyResult{
		Tests:        allTests,
		BindingSpecs: bindingSpecs,
	}, nil
}

// GeneratePairwiseTests generates EQUAL tests for consecutive pairs
// in the given list. E.g., (a b c) -> ((EQUAL a b) (EQUAL b c)).
func GeneratePairwiseTests(tests interface{}) interface{} {
	if tests == nil {
		return nil
	}
	cons, ok := tests.(*Cons)
	if !ok {
		return nil
	}
	if cons.Cdr == nil {
		return nil
	}
	rest, ok := cons.Cdr.(*Cons)
	if !ok {
		return nil
	}
	return &Cons{
		Car: SExprList(SymEqual, cons.Car, rest.Car),
		Cdr: GeneratePairwiseTests(cons.Cdr),
	}
}

// GenerateUnifyTests generates a list of explicit tests for matching
// the given pattern. Assumes that the pattern to be tested will be
// in variable "P". Tests are returned in backward order.
//
// Example:
//
//	(generate-unify-tests '(foo ?x) nil nil 'P)
//	  => ((NULL (CDR (CDR P)))
//	      (EQUAL ?X (CAR (CDR P)))
//	      (CONSP (CDR P))
//	      (EQUAL (QUOTE FOO) (CAR P))
//	      (CONSP P))
func GenerateUnifyTests(pattern interface{}, vars []string, tests interface{}, path interface{}) interface{} {
	if pattern == nil {
		// End of pattern: (null path)
		return &Cons{
			Car: SExprList(SymNull, path),
			Cdr: tests,
		}
	}

	if IsVariable(pattern) {
		s := pattern.(string)

		// Check if pattern is a member of vars.
		if stringMember(s, vars) {
			// Must see if the pattern has been bound elsewhere.
			previous := assocSExpr(s, tests)
			if previous != nil {
				// Add this position to test it:
				// push path onto (cdr previous)
				pushCdr(previous, path)
				return tests
			}
			// First occurrence: (pattern path)
			return &Cons{
				Car: SExprList(s, path),
				Cdr: tests,
			}
		}

		// Variable not in vars; it must be bound, so test against current value.
		// (equal pattern path)
		return &Cons{
			Car: SExprList(SymEqual, pattern, path),
			Cdr: tests,
		}
	}

	// Check for number.
	if isNumber(pattern) {
		// (and (numberp path) (= pattern path))
		return &Cons{
			Car: SExprList(SymAnd, SExprList(SymNumberp, path), SExprList(Symbol("="), pattern, path)),
			Cdr: tests,
		}
	}

	// Check for non-list atom.
	cons, ok := pattern.(*Cons)
	if !ok {
		// (equal (quote pattern) path)
		return &Cons{
			Car: SExprList(SymEqual, SExprList(SymQuote, pattern), path),
			Cdr: tests,
		}
	}

	// Recurse on a list.
	// First add (consp path) test, then recurse on car with (car path),
	// then recurse on cdr with (cdr path).
	innerTests := &Cons{
		Car: SExprList(SymConsp, path),
		Cdr: tests,
	}
	carPath := SExprList(SymCar, path)
	innerTests = toConsOrNil(GenerateUnifyTests(cons.Car, vars, innerTests, carPath))
	cdrPath := SExprList(SymCdr, path)
	return GenerateUnifyTests(cons.Cdr, vars, innerTests, cdrPath)
}

// --- Helper functions ---

// SExprList builds a proper Lisp list (a chain of Cons cells terminated by nil)
// from the given elements.
func SExprList(elems ...interface{}) *Cons {
	if len(elems) == 0 {
		return nil
	}
	result := &Cons{Car: elems[len(elems)-1], Cdr: nil}
	for i := len(elems) - 2; i >= 0; i-- {
		result = &Cons{Car: elems[i], Cdr: result}
	}
	return result
}

// SliceToSExpr converts a Go slice to a proper Lisp list.
func SliceToSExpr(elems []interface{}) interface{} {
	if len(elems) == 0 {
		return nil
	}
	var result interface{}
	for i := len(elems) - 1; i >= 0; i-- {
		result = &Cons{Car: elems[i], Cdr: result}
	}
	return result
}

// SExprToSlice converts a proper Lisp list to a Go slice.
func SExprToSlice(sexpr interface{}) []interface{} {
	var result []interface{}
	for sexpr != nil {
		cons, ok := sexpr.(*Cons)
		if !ok {
			// Dotted pair: add the final atom.
			result = append(result, sexpr)
			break
		}
		result = append(result, cons.Car)
		sexpr = cons.Cdr
	}
	return result
}

// Car returns the car of a Cons cell, or nil if not a Cons.
func Car(x interface{}) interface{} {
	if cons, ok := x.(*Cons); ok {
		return cons.Car
	}
	return nil
}

// Cdr returns the cdr of a Cons cell, or nil if not a Cons.
func Cdr(x interface{}) interface{} {
	if cons, ok := x.(*Cons); ok {
		return cons.Cdr
	}
	return nil
}

// Sublis substitutes variable bindings throughout an s-expression.
// Each binding maps a variable name to a replacement value.
func Sublis(alist []Binding, expr interface{}) interface{} {
	if expr == nil {
		return nil
	}
	if s, ok := expr.(string); ok {
		for _, b := range alist {
			if b.Var == s {
				return b.Val
			}
		}
		return expr
	}
	if sym, ok := expr.(Symbol); ok {
		for _, b := range alist {
			if b.Var == string(sym) {
				return b.Val
			}
		}
		return expr
	}
	cons, ok := expr.(*Cons)
	if !ok {
		return expr
	}
	newCar := Sublis(alist, cons.Car)
	newCdr := Sublis(alist, cons.Cdr)
	return &Cons{Car: newCar, Cdr: newCdr}
}

// stringMember returns true if s is in the given slice.
func stringMember(s string, slice []string) bool {
	for _, v := range slice {
		if v == s {
			return true
		}
	}
	return false
}

// assocSExpr looks up a key (string) in a Lisp alist (list of lists).
// Returns the matching cons cell or nil.
func assocSExpr(key string, alist interface{}) *Cons {
	for alist != nil {
		cons, ok := alist.(*Cons)
		if !ok {
			return nil
		}
		entry, ok := cons.Car.(*Cons)
		if ok {
			if s, ok := entry.Car.(string); ok && s == key {
				return entry
			}
		}
		alist = cons.Cdr
	}
	return nil
}

// pushCdr appends a value to the end of the cdr list of a Cons cell.
// This mutates the cons cell in place, mimicking Lisp's (push path (cdr previous)).
func pushCdr(cons *Cons, val interface{}) {
	// In the Lisp code, (push path (cdr previous)) prepends path to the cdr list.
	cons.Cdr = &Cons{Car: val, Cdr: cons.Cdr}
}

// isNumber returns true if x is a numeric type.
func isNumber(x interface{}) bool {
	switch x.(type) {
	case int, int8, int16, int32, int64,
		uint, uint8, uint16, uint32, uint64,
		float32, float64:
		return true
	default:
		return false
	}
}

// toConsOrNil converts an interface{} that should be a *Cons or nil to *Cons.
func toConsOrNil(x interface{}) *Cons {
	if x == nil {
		return nil
	}
	if c, ok := x.(*Cons); ok {
		return c
	}
	return nil
}

File: atms/java/atms/Funify.java
// -*- Mode: Java -*-
//
// Extra pattern-matching facilities for FTRE.
// Translated from funify.lisp, last edited 1/29/93 by KDF.
//
// Copyright (c) 1988-1992, Kenneth D. Forbus and Johan de Kleer.
// All rights reserved.
//
// See the file legal.txt for a paragraph stating scope of permission
// and disclaimer of warranty. The above copyright notice and that
// paragraph must be included in any separate copy of this file.

package atms;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import atms.Unify.Cons;

/**
 * Extra pattern-matching facilities for FTRE.
 *
 * <p>Provides compile-time-style helpers that manipulate symbolic
 * s-expression patterns at runtime.  The functions here build
 * explicit test and binding structures used by the rule compiler.
 *
 * <p>In the original Lisp code these are macro-expansion helpers;
 * here they are translated as regular static methods that operate
 * on {@link Cons}-based s-expressions.
 */
public class Funify {

    // ---------------------------------------------------------------
    // Thread-local *bound-vars* — the Lisp special variable.
    // ---------------------------------------------------------------

    /**
     * Corresponds to the Lisp special variable {@code *bound-vars*}.
     * Holds the list of variables that are considered bound in the
     * current expansion context.  Use {@link #getBoundVars()} and
     * {@link #setBoundVars(List)} to access it.
     */
    private static final ThreadLocal<List<String>> boundVars =
        ThreadLocal.withInitial(ArrayList::new);

    /** Get the current bound-vars list. */
    public static List<String> getBoundVars() {
        return boundVars.get();
    }

    /** Set the current bound-vars list. */
    public static void setBoundVars(List<String> vars) {
        boundVars.set(vars);
    }

    // ---------------------------------------------------------------
    // Result type for generate-match-body (multiple return values).
    // ---------------------------------------------------------------

    /**
     * Return type for {@link #generateMatchBody}, corresponding to
     * the Lisp {@code (values ...)} of that function.
     */
    public static class MatchBody {
        private final List<Object> tests;
        private final List<Object> bindingSpecs;

        public MatchBody(List<Object> tests, List<Object> bindingSpecs) {
            this.tests = tests;
            this.bindingSpecs = bindingSpecs;
        }

        /** The combined list of structure, equality, and extra tests. */
        public List<Object> getTests() { return tests; }

        /** The list of binding path specifications. */
        public List<Object> getBindingSpecs() { return bindingSpecs; }
    }

    // ---------------------------------------------------------------
    // quotize
    // ---------------------------------------------------------------

    /**
     * Translate a pattern into an s-expression that, when evaluated,
     * would reconstruct the pattern with variables left unquoted.
     *
     * <p>Corresponds to:
     * <pre>
     * (defun quotize (pattern) ...)
     * </pre>
     *
     * @param pattern an s-expression (possibly containing variables)
     * @return an s-expression with QUOTE/cons wrappers
     */
    public static Object quotize(Object pattern) {
        // (null pattern) => nil
        if (pattern == null) {
            return null;
        }
        // (variable? pattern) => pattern
        if (Unify.isVariable(pattern)) {
            return pattern;
        }
        // (not (listp pattern)) => (list 'QUOTE pattern)
        if (!(pattern instanceof Cons)) {
            return Unify.list("QUOTE", pattern);
        }
        Cons c = (Cons) pattern;
        // ((eq (car pattern) :EVAL) => (cadr pattern))
        if (":EVAL".equals(c.getCar()) || "EVAL".equals(c.getCar())) {
            // cadr = car of cdr
            return Unify.car(c.getCdr());
        }
        // (t `(cons ,(quotize (car pattern)) ,(quotize (cdr pattern))))
        return Unify.list("CONS",
                          quotize(c.getCar()),
                          quotize(c.getCdr()));
    }

    // ---------------------------------------------------------------
    // pattern-free-variables
    // ---------------------------------------------------------------

    /**
     * Return the list of free variables in {@code pattern}.
     * A variable is free if it is not in the previously-collected
     * list and not in the current {@link #getBoundVars()}.
     *
     * <p>Corresponds to:
     * <pre>
     * (defun pattern-free-variables (pattern) ...)
     * </pre>
     */
    public static List<String> patternFreeVariables(Object pattern) {
        return patternFreeVars1(pattern, new ArrayList<>());
    }

    /**
     * Recursive helper for {@link #patternFreeVariables}.
     *
     * <p>Corresponds to:
     * <pre>
     * (defun pattern-free-vars1 (pattern vars) ...)
     * </pre>
     */
    public static List<String> patternFreeVars1(Object pattern, List<String> vars) {
        if (pattern == null) {
            return vars;
        }
        if (Unify.isVariable(pattern)) {
            String v = (String) pattern;
            if (vars.contains(v) || getBoundVars().contains(v)) {
                return vars;
            }
            List<String> extended = new ArrayList<>(vars);
            extended.add(v);
            return extended;
        }
        if (!(pattern instanceof Cons)) {
            return vars;
        }
        Cons c = (Cons) pattern;
        return patternFreeVars1(c.getCdr(),
                                patternFreeVars1(c.getCar(), vars));
    }

    // ---------------------------------------------------------------
    // generate-match-body
    // ---------------------------------------------------------------

    /**
     * Generate an explicit match body consisting of tests and
     * binding specifications for the given pattern.
     *
     * <p>Corresponds to:
     * <pre>
     * (defun generate-match-body (pattern vars extra-test ...) ...)
     * </pre>
     *
     * @param pattern   the s-expression pattern
     * @param vars      the list of known pattern variables
     * @param extraTest an optional extra test expression (may be null)
     * @return a {@link MatchBody} containing tests and binding specs
     */
    public static MatchBody generateMatchBody(Object pattern,
                                              List<String> vars,
                                              Object extraTest) {
        List<Object> structureTests = new ArrayList<>();
        List<Object> equalTests = new ArrayList<>();
        List<Object> bindingSpecs = new ArrayList<>();
        // var-alist maps variable names to their last path occurrence.
        List<Unify.Binding> varAlist = new ArrayList<>();

        List<Object> rawTests = generateUnifyTests(pattern, vars,
                                                   new ArrayList<>(), "P");
        for (Object testObj : rawTests) {
            if (!(testObj instanceof Cons)) {
                structureTests.add(testObj);
                continue;
            }
            Cons test = (Cons) testObj;
            Object head = test.getCar();
            if (Unify.isVariable(head)) {
                // test looks like (?x path1 path2 ...)
                List<Object> paths = consToList(test.getCdr());
                equalTests.addAll(generatePairwiseTests(paths));
                if (extraTest != null && !paths.isEmpty()) {
                    varAlist.add(new Unify.Binding((String) head,
                                                   paths.get(paths.size() - 1)));
                }
                if (!paths.isEmpty()) {
                    bindingSpecs.add(paths.get(paths.size() - 1));
                }
            } else {
                structureTests.add(testObj);
            }
        }

        // (setq extra-test (sublis var-alist extra-test))
        extraTest = sublis(varAlist, extraTest);

        // (when (pattern-free-variables extra-test) (error ...))
        if (extraTest != null && !patternFreeVariables(extraTest).isEmpty()) {
            throw new RuntimeException(
                "Rule test includes free variable: " + extraTest);
        }

        List<Object> allTests = new ArrayList<>(structureTests);
        allTests.addAll(equalTests);
        if (extraTest != null) {
            allTests.add(extraTest);
        }

        return new MatchBody(allTests, bindingSpecs);
    }

    // ---------------------------------------------------------------
    // generate-pairwise-tests
    // ---------------------------------------------------------------

    /**
     * Given a list of path expressions, generate pairwise EQUAL
     * tests between consecutive elements.
     *
     * <p>Corresponds to:
     * <pre>
     * (defun generate-pairwise-tests (tests) ...)
     * </pre>
     */
    public static List<Object> generatePairwiseTests(List<Object> tests) {
        List<Object> result = new ArrayList<>();
        for (int i = 0; i + 1 < tests.size(); i++) {
            result.add(Unify.list("EQUAL", tests.get(i), tests.get(i + 1)));
        }
        return result;
    }

    // ---------------------------------------------------------------
    // generate-unify-tests
    // ---------------------------------------------------------------

    /**
     * Generate a list of explicit tests for matching the given
     * pattern.  Assumes the pattern to be tested is in variable
     * {@code "P"}.  Tests are returned in backward order.
     *
     * <p>Example (Lisp notation):
     * <pre>
     * (generate-unify-tests '(foo ?x) nil nil 'P)
     *  => ((NULL (CDR (CDR P)))
     *      (EQUAL ?X (CAR (CDR P)))
     *      (CONSP (CDR P))
     *      (EQUAL (QUOTE FOO) (CAR P))
     *      (CONSP P))
     * </pre>
     *
     * <p>Corresponds to:
     * <pre>
     * (defun generate-unify-tests (pattern vars tests path) ...)
     * </pre>
     *
     * @param pattern the s-expression pattern
     * @param vars    known pattern variables (for multi-occurrence tracking)
     * @param tests   accumulator of tests so far
     * @param path    current accessor path expression
     * @return extended list of tests
     */
    public static List<Object> generateUnifyTests(Object pattern,
                                                  List<String> vars,
                                                  List<Object> tests,
                                                  Object path) {
        // (null pattern) => (cons `(null ,path) tests)
        if (pattern == null) {
            List<Object> result = new ArrayList<>(tests);
            result.add(Unify.list("NULL", path));
            return result;
        }

        // (member pattern vars)
        if (pattern instanceof String && vars.contains(pattern)) {
            // Look for a previous entry for this variable in tests.
            for (Object t : tests) {
                if (t instanceof Cons) {
                    Cons tc = (Cons) t;
                    if (Objects.equals(tc.getCar(), pattern)) {
                        // push path onto (cdr previous)
                        // We need to mutate the list in-place like the Lisp version.
                        // Since Cons is immutable, we rebuild the tests list.
                        List<Object> result = new ArrayList<>();
                        for (Object t2 : tests) {
                            if (t2 == t) {
                                // Append path to this entry's cdr list.
                                result.add(appendToCons((Cons) t2, path));
                            } else {
                                result.add(t2);
                            }
                        }
                        return result;
                    }
                }
            }
            // Not found — add a new entry.
            List<Object> result = new ArrayList<>(tests);
            result.add(Unify.list(pattern, path));
            return result;
        }

        // (variable? pattern) => (cons `(equal ,pattern ,path) tests)
        if (Unify.isVariable(pattern)) {
            List<Object> result = new ArrayList<>(tests);
            result.add(Unify.list("EQUAL", pattern, path));
            return result;
        }

        // (numberp pattern) => (cons `(and (numberp ,path) (= ,pattern ,path)) tests)
        if (pattern instanceof Number) {
            List<Object> result = new ArrayList<>(tests);
            result.add(Unify.list("AND",
                                  Unify.list("NUMBERP", path),
                                  Unify.list("=", pattern, path)));
            return result;
        }

        // (atom pattern) => (cons `(equal ',pattern ,path) tests)
        if (!(pattern instanceof Cons)) {
            List<Object> result = new ArrayList<>(tests);
            result.add(Unify.list("EQUAL",
                                  Unify.list("QUOTE", pattern),
                                  path));
            return result;
        }

        // Recurse on a list.
        Cons c = (Cons) pattern;
        // First add (consp path) test, then recurse on car, then cdr.
        List<Object> withConsp = new ArrayList<>(tests);
        withConsp.add(Unify.list("CONSP", path));

        List<Object> afterCar = generateUnifyTests(
            c.getCar(), vars, withConsp,
            Unify.list("CAR", path));

        return generateUnifyTests(
            c.getCdr(), vars, afterCar,
            Unify.list("CDR", path));
    }

    // ---------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------

    /**
     * Substitute variable bindings throughout an s-expression.
     * Corresponds to the Lisp {@code sublis} function.
     *
     * @param alist    list of bindings (variable -> replacement)
     * @param expr     the expression to transform
     * @return the expression with substitutions applied
     */
    public static Object sublis(List<Unify.Binding> alist, Object expr) {
        if (expr == null || alist == null || alist.isEmpty()) {
            return expr;
        }
        // Check if expr itself is a key in the alist.
        for (Unify.Binding b : alist) {
            if (Objects.equals(b.getVar(), expr)) {
                return b.getVal();
            }
        }
        if (expr instanceof Cons) {
            Cons c = (Cons) expr;
            Object newCar = sublis(alist, c.getCar());
            Object newCdr = sublis(alist, c.getCdr());
            if (newCar == c.getCar() && newCdr == c.getCdr()) {
                return expr; // No change — preserve identity.
            }
            return new Cons(newCar, newCdr);
        }
        return expr;
    }

    /**
     * Convert a Cons-based proper list to a {@code List<Object>}.
     */
    public static List<Object> consToList(Object sexpr) {
        List<Object> result = new ArrayList<>();
        Object cur = sexpr;
        while (cur instanceof Cons) {
            result.add(((Cons) cur).getCar());
            cur = ((Cons) cur).getCdr();
        }
        // If cur is non-null here it was a dotted pair; ignore tail.
        return result;
    }

    /**
     * Append an element to the end of a Cons-based proper list,
     * returning a new Cons list.  Used to simulate the Lisp
     * {@code (push path (cdr previous))} which mutably extends
     * the cdr-chain of a test entry.
     */
    private static Cons appendToCons(Cons list, Object element) {
        List<Object> items = consToList(list);
        items.add(element);
        return (Cons) Unify.list(items.toArray());
    }
}
